File: leet_LIST_0086_partition.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function partition(head: ListNode | null, x: number): ListNode | null {
    if (head == null) {
        return head
    }
    let L: number[] = []
    let R: number[] = []
    let p: ListNode
    let n: number

    p = head
    while (p) {
        n = p.val
        if (n < x) {
            L.push(n)
        } else {
            R.push(n)
        }
        p = p.next
    }

    p = head
    for (let i of L) {
        p.val = i
        p = p.next
    }
    for (let i of R) {
        p.val = i
        p = p.next
    }
    return head
};

File: leet_LIST_0086_partition.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
    if head == nil {
        return head
    }

    var L, R []int
    p := head

    for p != nil {
        n := p.Val
        if n < x {
            L = append(L, n)
        } else {
            R = append(R, n)
        }
        p = p.Next
    }

    p = head
    for _, n := range L {
        p.Val = n
        p = p.Next
    }

    for _, n := range R {
        p.Val = n
        p = p.Next
    }

    return head
}
